rpcs3/rpcs3qt/main_window.cpp

```diff
@@ -884,56 +916,83 @@ void main_window::HandlePupInstallation(QString file_path)
 				if (dev_flash_tar_f.size() < 3)
 				{
 					gui_log.error("Error while installing firmware: PUP contents are invalid.");
-					QMessageBox::critical(this, tr("Firmware Installation Failed"), tr("Firmware installation failed: Firmware could not be decompressed"));
-					progress = -1;
+					critical(tr("Firmware installation failed: Firmware could not be decompressed"));
+					progress = UINT32_MAX;
+					break;
 				}
 
 				tar_object dev_flash_tar(dev_flash_tar_f[2]);
 				if (!dev_flash_tar.extract(g_cfg.vfs.get_dev_flash(), "dev_flash/"))
 				{
 					gui_log.error("Error while installing firmware: TAR contents are invalid.");
-					QMessageBox::critical(this, tr("Firmware Installation Failed"), tr("Firmware installation failed: Firmware contents could not be extracted."));
-					progress = -1;
+					critical(tr("Firmware installation failed: Firmware contents could not be extracted."));
+					progress = UINT32_MAX;
+					break;
 				}
 
-				if (progress >= 0)
-					progress += 1;
+				if (!progress.try_inc(static_cast<u32>(update_filenames.size())))
+				{
+					break;
+				}
 			}
 		});
 
+		if (!mw)
+		{
+			// Join thread
+			break;
+		}
+
+		progress_dialog pdlg(tr("RPCS3 Firmware Installer"), tr("Installing firmware version %1\nPlease wait...").arg(qstr(version_string)), tr("Cancel"), 0, static_cast<int>(update_filenames.size()), false, mw);
+		pdlg.show();
+
 		// Wait for the completion
-		while (std::this_thread::sleep_for(5ms), std::abs(progress) < pdlg.maximum())
+		for (uint current = progress; current < update_filenames.size(); std::this_thread::sleep_for(5ms), current = progress)
 		{
+			// Update progress window
+			pdlg.SetValue(current);
+			QCoreApplication::processEvents();
+
 			if (pdlg.wasCanceled())
 			{
-				progress = -1;
+				progress = UINT32_MAX;
 				break;
 			}
-			// Update progress window
-			pdlg.SetValue(static_cast<int>(progress));
-			QCoreApplication::processEvents();
 		}
 
 		update_files_f.close();
 		pup_f.close();
 
-		if (progress > 0)
+		if (progress == update_filenames.size())
 		{
 			pdlg.SetValue(pdlg.maximum());
 			std::this_thread::sleep_for(100ms);
 		}
 	}
+	while (false);
 
 	// Update with newly installed PS3 fonts
 	Q_EMIT RequestGlobalStylesheetChange();
 
-	if (progress > 0)
+	if (progress != update_filenames.size())
 	{
-		gui_log.success("Successfully installed PS3 firmware version %s.", version_string);
-		m_gui_settings->ShowInfoBox(tr("Success!"), tr("Successfully installed PS3 firmware and LLE Modules!"), gui::ib_pup_success, this);
+		return false;
+	}
+
+	QString success_str = tr("Successfully installed PS3 firmware (v%1) and LLE Modules!").arg(qstr(version_string));
```

Comment:
we don't translate log messages, as these are supposed to be readable by developers.
So just copy paste the message without translation to the log message

Reply:
Removed.